trin-core/src/portalnet/storage.rs

```diff
@@ -0,0 +1,605 @@
+        id_as_u256.to_big_endian(y);
+
+        y.clone()
+
+    }
+    
+    
+    pub fn should_store(&self, key: &String) -> bool {
+
+        let content_id = self.content_key_to_content_id(key);
+
+        // Don't store if we already have the data
+        match self.db.get(&content_id) {
+            Ok(Some(_)) => {
+                return false;
+            },
+            Err(e) => panic!("Unable to respond to FindContent: {}", e),
+            _ => ()
+        }
+
+        // Don't store if it's outside our radius
+        if self.data_radius < u64::MAX {
+            self.distance_to_content_id(&content_id) < self.data_radius
+        } else {
+            true
+        }
+
+    }
+
+    fn meta_db_insert(&self, content_id: &[u8; 32], content_key: &String, value: &String) {
+
+        let content_id_as_u32: u32 = PortalStorage::byte_vector_to_u32(content_id.clone().to_vec());
+
+        let value_size = value.len();
+
+        self.meta_db.execute(
+            INSERT_QUERY,
+            params![content_id.to_vec(), content_id_as_u32, content_key, value_size],
+        ).unwrap();
+
+    }
+
+    fn meta_db_remove(&self, content_id: &[u8; 32]) {
+
+        self.meta_db.execute(
+            DELETE_QUERY,
+            [content_id.to_vec()],
+        ).unwrap();
+        
+    }
+
+    pub fn store(&mut self, key: &String, value: &String) {
+
+        // Check whether we should store this data
+        if !self.should_store(&key) {
+            println!("Not storing: {}", key);
+            return;
+        }   
+
+        let content_id = self.content_key_to_content_id(key);
+
+        // Store the data 
+        self.db.put(&content_id, value).expect("Failed to write to DB");
+        self.meta_db_insert(&content_id, &key, value);
+
+        // Update the farthest key if this key is either 1.) the first key ever or 2.) farther than the current farthest
+        match self.farthest_content_id.as_ref() {
+            None => {
+                self.farthest_content_id = Some(content_id);
+            },
+            Some(farthest) => {
+                if self.distance_to_content_id(&content_id) > self.distance_to_content_id(&farthest) {
+                    self.farthest_content_id = Some(content_id.clone());
+                }
+            }
+        }
+
+        // Delete furthest data until our data usage is less than capacity
+        while self.capacity_reached() {
+
+            let id_to_remove = self.farthest_content_id.clone();
+            
+            println!("Capacity reached, deleting farthest: {}", hex::encode(&id_to_remove.clone().unwrap()));
+
+            self.db.delete(&id_to_remove.clone().unwrap()).expect("Failed to delete key.");
+            self.meta_db_remove(&id_to_remove.clone().unwrap());
+
+            self.db.flush().expect("Failed to flush db.");
+            
+            match self.find_farthest_content_id() {
+                None => { error!("Failed to find farthest!") },
```

Comment:
```suggestion
                None => {
                    error!("Database is over-capacity, but could not find find entry to delete!");
                    self.farthest_content_id = None;
                    // stop attempting to delete, to avoid risk of infinite loop
                    break;
                },
```